### ISSUE TITLE
Segfault when entering scheme context manager in recent versions

### ISSUE THREAD

**Issue by titodalcanton** [gwastro/pycbc#4076]
@soumenroy pointed me to the fact that `pycbc_banksim` is segfaulting when getting to the `with ctx` line. Indeed, I can reproduce the segfault using PyCBC 2.0.4 and this minimal script:
```python
from argparse import ArgumentParser
import pycbc.scheme


parser = ArgumentParser(description=__doc__)
pycbc.scheme.insert_processing_option_group(parser)
options = parser.parse_args()

ctx = pycbc.scheme.from_cli(options)

with ctx:
    pass
```
However I cannot reproduce the crash with PyCBC 2.0.3 (using the same virtualenv). So I suspect this is somehow a consequence of #4054. Investigating.

**Comment by titodalcanton** [comment]
@josh-willis 
```
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py 
Segmentation fault
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py --processing-scheme mkl
Segmentation fault
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py --processing-scheme cpu
Segmentation fault
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py --processing-scheme cuda
Traceback (most recent call last):
  File "/home/tito.canton/segfault.py", line 9, in <module>
    ctx = pycbc.scheme.from_cli(options)
  File "/home/tito.canton/virtualenvs/misc_py39/lib/python3.9/site-packages/pycbc/scheme.py", line 280, in from_cli
    ctx = CUDAScheme(opt.processing_device_id)
  File "/home/tito.canton/virtualenvs/misc_py39/lib/python3.9/site-packages/pycbc/scheme.py", line 115, in __init__
    raise RuntimeError("Install PyCUDA to use CUDA processing")
RuntimeError: Install PyCUDA to use CUDA processing
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py --processing-scheme gpu
Segmentation fault
(misc_py39) [tito.canton@ldas-pcdev1 ~]$ python segfault.py --processing-scheme numpy
Segmentation fault
```

**Comment by josh-willis** [comment]
@titodalcanton I don't know off hand, and am about to be on vacation for a couple of weeks, so I will not be able to move forward debugging and fixing this, though obviously it's urgent.  My immediate suggestion would be that based on the discussion in #4054 , we may want to go back to having `RTLD_GLOBAL` be the default mode for libraries, and then for things like FFTW explicitly set `RTLD_DEEPBIND`.  The logic to see if `os.RTLD_DEEPBIND` exists on a given system should presumably still be housed in `pycbc.libutils`.  But all of that is just what I'd try first; I don't actually know that there is something that will fix this without triggering the problem that #4054 addressed.

Also, we should very carefully look at our test suite, it definitely should have been catching this error, though given what you observed, it could have something to do with whether one is in a virtual environment as opposed to anything else (system install, container, conda?). @ahnitz any ideas?

**Comment by titodalcanton** [comment]
I am starting to suspect that this happens when I use a virtualenv derived from the Python in the igwn-py39 Conda env. I can reproduce this on Hawk with a virtualenv created from igwn-py39.

**Comment by josh-willis** [comment]
@titodalcanton So you may well be right about where this is occurring, though I'm still not sure why it happens there (in some virtualenvs).

Looking at your minimal reproducing case, it only relies on the `scheme` package, and clearly segfaults for any of the CPU-based schemes (note that `gpu` is not a valid scheme, so I think in that case you're actually just falling back to the default cpu scheme, though I thought it should raise an error in that case).

And after looking through [scheme.py](https://github.com/gwastro/pycbc/blob/master/pycbc/scheme.py), the only think I can think of is that the new `DEEPBIND` mode is not compatible with opening `libgomp`, which seems to be the only place that any of the library stuff that was changed in #4054 would interact with the scheme. I'm also not sure why we are still only opening `libgomp`; it's gnu specific and I would think if MKL is the scheme we we would want `libiomp` (IIRC).  But that's (probably) a separate issue.

So it might be worth trying whether changing `scheme.py` to open libgomp with the old default fixes this.  Though if it does, we should probably also revisit making `DEEPBIND` the new default in `libutils.py`, and keep `GLOBAL` as we had before, using `DEEPBIND` only explicitly where we must.

**Comment by spxiwh** [comment]
A much more shortened example to reproduce this (which removes PyCBC from the loop):

```
import ctypes

_libgomp = ctypes.CDLL('libgomp.so.1')

_libgomp.omp_set_num_threads(2)
```

**Comment by titodalcanton** [comment]
As far as I can tell, this has been fixed by #4079 for me, so I am closing for now.
